lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,471 @@
+        )
+        self.framework.observe(
+            self.charm.on[TLS_CERTIFICATE_ACCESS_RELATION_NAME].relation_broken,
+            self._on_tls_access_certificate_relation_broken,
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_available, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_expiring, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_invalidated, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.charm.on[SEND_CA_CERT_RELATION_NAME].relation_joined,
+            self._on_certificate_config_changed,
+        )
+
+    def _on_certificate_config_changed(self, event: EventBase):
+        """Handle TLS configuration changes. Makes the charm reconfigure its environment."""
+        self.charm.on.config_changed.emit()
+
+    def _on_tls_access_certificate_relation_broken(self, event: RelationBrokenEvent):
+        """Handles leaving the tls access relation.
+
+        Regenerates self signed certificates from the saved self generated root CA
+        and reloads vault.
+        """
+        if not self.charm._ingress_address:
+            self._remove_all_certs_from_workload()
+            return
+        self._generate_self_signed_certs(self.charm._ingress_address)
+        logger.info(
+            "TLS: Saved Vault generated CA and self signed certificate to %s.",
+            self.charm.unit.name,
+        )
+        self._reload_vault_container()
+
+    def configure_certificates(self, ingress_address: str) -> None:
+        """Configures the certificates that are used to connect to and communicate with Vault.
+
+        Args:
+            ingress_address: The ip address for which the certificates will be configured for.
+        """
+        if self.charm.unit.is_leader() and not self._ca_certificate_set_in_peer_relation():
+            ca_private_key, ca_certificate = generate_vault_ca_certificate()
+            self._set_ca_certificate_secret_in_peer_relation(ca_private_key, ca_certificate)
+            logger.info("TLS: Saved the Vault generated CA cert in juju secrets.")
+
+        if self.charm.model.get_relation(TLS_CERTIFICATE_ACCESS_RELATION_NAME):
+            if self._should_request_new_certificate():
+                self._send_new_certificate_request_to_provider(
+                    self._pull_tls_file_from_workload(File.CSR), ingress_address
+                )
+                logger.info("TLS: CSR for unit %s sent to access relation.", self.charm.unit.name)
+            if (
+                assigned_cert := self.tls_access._find_certificate_in_relation_data(
+                    self._pull_tls_file_from_workload(File.CSR)
+                )
+            ) and assigned_cert["certificate"] != self._pull_tls_file_from_workload(File.CERT):
```

Comment:
I think doing this in multiple statement would be a bit more readable:

```suggestion
            assigned_cert = self.tls_access._find_certificate_in_relation_data(
                                      self._pull_tls_file_from_workload(File.CSR))
            if assigned_cert and assigned_cert["certificate"] != self._pull_tls_file_from_workload(File.CERT):
```

Reply:
Done.